contracts/eosio.system/src/finalizer_key.cpp

```diff
@@ -0,0 +1,322 @@
+   * @post is_savanna_consensus returns true
+   */
+   void system_contract::switchtosvnn() {
+      require_auth(get_self());
+
+      check(!is_savanna_consensus(), "switchtosvnn can be run only once");
+
+      using value_type = std::pair<eosio::producer_authority, uint16_t>;
+      std::vector<value_type> top_producers;
+      std::vector<eosio::finalizer_authority> finalizer_authorities;
+      std::vector<uint64_t> first_finalizer_key_set;
+      top_producers.reserve(21);
+      finalizer_authorities.reserve(21);
+      first_finalizer_key_set.reserve(21);
+
+      // From up to 30 top producers, find 21 producers that meet all the normal requirements
+      // for being a proposer and also have an active finalizer key
+      uint32_t i = 0;
+      auto idx = _producers.get_index<"prototalvote"_n>();
+      for( auto it = idx.cbegin(); it != idx.cend() && top_producers.size() < 21 && 0 < it->total_votes && it->active() && i < 30; ++it, ++i ) {
+         auto finalizer = _finalizers.find( it->owner.value );
+         if( finalizer == _finalizers.end() ) {
+            // The producer is not in finalizers table, indicating it does not have an
+            // active registered finalizer key. Try next one.
+            continue;
+         }
+
+         // This should never happen. Double check the finalizer has an active key just in case
+         if( finalizer->active_finalizer_key_binary.empty() ) {
+            continue;
+         }
+
+         // builds up producer_authority
+         top_producers.emplace_back(
+            eosio::producer_authority{
+               .producer_name = it->owner,
+               .authority     = it->get_producer_authority()
+            },
+            it->location
+         );
+
+         // stores finalizer_key_id
+         first_finalizer_key_set.emplace_back(finalizer->active_finalizer_key_id);
+
+         // builds up finalizer_authorities
+         finalizer_authorities.emplace_back(
+            eosio::finalizer_authority{
+               .description = finalizer->finalizer_name.to_string(),
+               .weight      = 1,
+               .public_key  = finalizer->active_finalizer_key_binary
+            }
+         );
+      }
+
+      check( top_producers.size() > 0 && top_producers.size() >= _gstate.last_producer_schedule_size, "not enough top producers have registered finalizer keys" );
+
+      std::sort( top_producers.begin(), top_producers.end(), []( const value_type& lhs, const value_type& rhs ) {
+         return lhs.first.producer_name < rhs.first.producer_name; // sort by producer name
+         // return lhs.second < rhs.second; // sort by location
+      } );
+
+      std::vector<eosio::producer_authority> producers;
+
+      producers.reserve(top_producers.size());
+      for( auto& item : top_producers )
+         producers.push_back( std::move(item.first) );
+
+      // Should set_finalizers() be called before set_proposed_producers()
+      // during transition?
```

Comment:
Doesn't matter. `set_proposed_producers` is going to be ignored. Probably should not even bother calling it below. Which means you don't need the sort above or `producers` or `top_producers` vectors.